### PR TITLE
`MangleNames`: Create Haskell names for implicit fields based on external types

### LINKED ISSUE #2159: Create Haskell names for implicit fields based on external types (if any)
For C code like:

```c
struct S {
  struct { // Point
    int x; // X coordinate
    int y; // Y coordinate
  };
};
```

where the anonymous struct is specified using an external type `M.Point` using a binding spec, we generate:

```hs
data S = S { s_anon'x :: M.Point }
```

Currently the Haskell name of the field (`s_anon'x`) is based on a C field name that `hs-bindgen` synthesises for implicit fields. We could use the name of the external type to inform how we create the Haskell name. We could then generate, for example:

```hs
data S = S { s_point :: M.Point }
```

### PR DESCRIPTION
**Pull request by jorisdral**
Resolves #2159

---

- [ ] I have updated the changelog. I have included a migration hint if this is a breaking change.
- [ ] I have updated the manual.
- [ ] I have removed all mentions of closed tickets in the code.
- [ ] I have associated each new TODO item with a ticket

### CODE REVIEW

File: hs-bindgen/test-artefacts/headers/golden/types/anonymous/edge-cases/name_conflict.h [hunk 1]
```diff
@@ -0,0 +1,11 @@
+struct Rectangle {
+  struct { // Point2D
+      int x1; // top-left X coordinate
+      int y1; // top-left Y coordinate
+    };
+  struct { // Point3D
```
Comment (jorisdral) [inline hs-bindgen/test-artefacts/headers/golden/types/anonymous/edge-cases/name_conflict.h:6]:
```suggestion
  struct { // Point2D
```

File: hs-bindgen/src-internal/HsBindgen/Frontend/Pass/MangleNames/CreateNames.hs [hunk 2]
```diff
@@ -528,6 +528,29 @@ createFieldName hsName fieldCName = do
...
+          candidate = case strategy of
+            AddFieldPrefixes  -> hsName <> "_" <> extHsName.name.text
+            OmitFieldPrefixes -> fieldCName.text
+      name <- mkName (Proxy @Hs.NsVar) candidate
+      let scopedNamePair = ScopedNamePair{
+          cName  = fieldCName
+        , hsName = Hs.demoteNs name
+        }
+      modify $ #scopedNames %~ (scopedNamePair:)
+      pure scopedNamePair
```
Comment (jorisdral) [inline hs-bindgen/src-internal/HsBindgen/Frontend/Pass/MangleNames/CreateNames.hs:552]: reduce duplication with `createFieldName`

File: hs-bindgen/test-artefacts/fixtures/types/anonymous/edge-cases/name_conflict/bindingspec.yaml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  cname: struct @Rectangle_anon'x2
+  hsname: Rectangle_anon'x2
+hstypes:
+- hsname: Rectangle
+  representation:
+    record:
+      constructor: Rectangle
+      fields:
+      - rectangle_anon'x1
+      - rectangle_anon'x2
```
Comment (jorisdral) [inline hs-bindgen/test-artefacts/fixtures/types/anonymous/edge-cases/name_conflict/bindingspec.yaml:22]: These are the C names. They should be:
```suggestion
      - rectangle_Point2D
      - rectangle_Point2D
```

**Comment by edsko** [comment]
Going to close this as wontfix: while type names can be based on field names (because C ensures field names are (locally) unique), field names cannot be based on type names because type certainly _aren't_ unique.
